contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,440 @@
+    function _D(uint256 xWad, uint256 yWad) internal view returns (uint256) {
+        return Math.mulDiv(xWad, P, wad) + yWad;
+    }
+
+    /// @notice Amount of token in wad
+    /// @param dx Amount of token in token decimals
+    /// @param isToken0 Whether token is token0
+    function _amountWad(uint256 dx, bool isToken0) internal view returns (uint256) {
+        uint256 factor = isToken0 ? 10**(token0Decimals) : 10**(token1Decimals);
+        return Math.mulDiv(dx, wad, factor);
+    }
+
+    /// @notice Amount of token in token decimals
+    /// @param amount Amount of token in wad
+    /// @param isToken0 Whether token is token0
+    function _amountDecimals(uint256 amount, bool isToken0) internal view returns (uint256) {
+        uint256 factor = isToken0 ? 10**(token0Decimals) : 10**(token1Decimals);
+        return Math.mulDiv(amount, factor, wad);
+    }
+
+    /// @notice Calculates amount of tokens received on swap
+    /// @dev Returns zero if pool balances exceeded on swap or invalid inputs
+    /// @param tokenIndexFrom The index of the token in
+    /// @param tokenIndexTo The index of the token out
+    /// @param dx The amount of token in in token decimals
+    // TODO: test with protocol fees
+    function _calculateSwap(
+        uint8 tokenIndexFrom,
+        uint8 tokenIndexTo,
+        uint256 dx
+    )
+        internal
+        view
+        returns (
+            uint256 dy_,
+            uint256 dyFee_,
+            uint256 dyAdminFee_
+        )
+    {
+        if (tokenIndexFrom > 1 || tokenIndexTo > 1 || tokenIndexFrom == tokenIndexTo) return (0, 0, 0);
+
+        // get current token balances in pool adjusted for protocol fees
+        uint256 xWad = _amountWad(IERC20(token0).balanceOf(address(this)) - protocolFees0, true);
+        uint256 yWad = _amountWad(IERC20(token1).balanceOf(address(this)) - protocolFees1, false);
+
+        // convert to an amount in wad
+        uint256 amountInWad = _amountWad(dx, tokenIndexFrom == 0);
+
+        // calculate swap amount out wad
+        // @dev obeys P * x + y = D
+        uint256 amountOutWad;
+        if (tokenIndexFrom == 0) {
+            // get D balance before swap
+            uint256 _d = _D(xWad, yWad);
+
+            // token0 in for token1 out
+            xWad += amountInWad;
+
+            // check amount out won't exceed pool balance
+            uint256 prod = Math.mulDiv(P, xWad, wad);
+            if (_d < prod) return (0, 0, 0);
+
+            uint256 yWadAfter = _d - prod;
+            amountOutWad = yWad - yWadAfter;
+        } else {
+            // get D balance before swap
+            uint256 _d = _D(xWad, yWad);
+
+            // token1 in for token0 out
+            yWad += amountInWad;
+
+            // check amount out won't exceed pool balance
+            if (_d < yWad) return (0, 0, 0);
+
+            uint256 xWadAfter = Math.mulDiv(_d - yWad, wad, P);
+            amountOutWad = xWad - xWadAfter;
```

Comment:
Likewise, this is just `amountOutWad = Math.mulDiv(amountInWad, wad, P)` + LP balance check.